File: cpp/SAMImgEncoder.cpp
#include "SAMImgEncoder.h"
#include <onnxruntime_cxx_api.h>
#include <iostream>
#include <array>

SAMImgEncoder::SAMImgEncoder() :
	img_size(1024), qt(false)
{
}

cv::Size SAMImgEncoder::get_preprocess_shape(int oldh, int oldw, int long_side_length) {
    float scale = static_cast<float>(long_side_length) / std::max(oldh, oldw);
    int newh = static_cast<int>(oldh * scale + 0.5);
    int neww = static_cast<int>(oldw * scale + 0.5);
    return cv::Size(neww, newh);
}

cv::Mat SAMImgEncoder::apply_image(const cv::Mat& image) {

    cv::Mat resized_image, padded_image, chw_image;

    cv::Size target_size = get_preprocess_shape(image.rows, image.cols, img_size);
    //std::cout << "apply_image.target_size: " << target_size << std::endl; // w:1024, h:683
    
    // resize, interpolation=cv::INTER_LINEAR
    cv::resize(image, resized_image, target_size, 0, 0, cv::INTER_LINEAR);
    // debug:print resized_image size
    std::cout << "apply_image.resized_image.size: " << resized_image.size() << std::endl; //[1024 x 683]
    
    // Normalization per channel, pixel_mean({ 123.675, 116.28, 103.53 }), pixel_std({ 58.395, 57.12, 57.375 })
    cv::Mat channels[3];
    cv::split(resized_image, channels);
    channels[0] = (channels[0] - 123.675) / 58.395;
    channels[1] = (channels[1] - 116.28) / 57.12;
    channels[2] = (channels[2] - 103.53) / 57.375;
    cv::merge(channels, 3, resized_image);
    //debug: convert to vector, and print first 30 elements of vector.
    std::vector<float> resized_image_vec;
    resized_image_vec.assign((float*)resized_image.data, (float*)resized_image.data + resized_image.total() * resized_image.channels());
    std::cout << "apply_image.after_norm" << std::endl;
    for (int i = 0; i < 30; i++) {
		std::cout << resized_image_vec[i] << " ";
	}

    
    //pad to square
    int h = resized_image.rows;
    int w = resized_image.cols;
    int pad_h = img_size - h;
    int pad_w = img_size - w;
    assert(pad_h >= 0 && pad_w >= 0);
    cv::copyMakeBorder(resized_image, padded_image, 0, pad_h, 0, pad_w, cv::BORDER_CONSTANT, cv::Scalar(0, 0, 0));

    //HWC to CHW
    cv::dnn::blobFromImage(padded_image, chw_image);

    return chw_image;

}


std::vector<float> SAMImgEncoder::encode_img(const cv::Mat& x) {
	// apply image: 1. resize, 2. normalize, 3. pad to square, 4. HWC to CHW
    cv::Mat x_img = apply_image(x);
    
    // copy x_img to vector
    std::vector<float> x_img_vec;
    x_img_vec.assign((float*)x_img.data, (float*)x_img.data + x_img.total() * x_img.channels());

    // set input, output array
    constexpr int64_t numInputElements = 3 * 1024 * 1024;
    constexpr int64_t numOutputElements = 256 * 64 * 64;
    std::vector<float> input(numInputElements);
    std::vector<float> output(numOutputElements);

    // set shape
    const std::array<int64_t, 4> inputShape = { 1, 3, 1024, 1024 };
    const std::array<int64_t, 4> outputShape = { 1, 256, 64, 64 };
	
    // paths
    auto img_encoder_path = L"F:\\Hacarus\\SAM_zhu\\savedmodel\\ImageEncoderViT.onnx";
    auto img_encoder_path_qt = L"F:\\Hacarus\\SAM_zhu\\savedmodel\\ImageEncoderViT_qt.onnx";

    Ort::Env env; // create environment
    Ort::RunOptions runOptions; // set run options if needed
    Ort::Session session(nullptr); //

    // set sesson
    Ort::SessionOptions ort_session_options;
    session = Ort::Session(env, img_encoder_path, Ort::SessionOptions{ nullptr });

    // define Tensor
    auto memory_info = Ort::MemoryInfo::CreateCpu(OrtDeviceAllocator, OrtMemTypeCPU);
    auto inputTensor = Ort::Value::CreateTensor<float>(memory_info, input.data(), input.size(), inputShape.data(), inputShape.size());
    auto outputTensor = Ort::Value::CreateTensor<float>(memory_info, output.data(), output.size(), outputShape.data(), outputShape.size());

    // copy image data to input array
    std::copy(x_img_vec.begin(), x_img_vec.end(), input.begin());


    // define names
    Ort::AllocatorWithDefaultOptions ort_alloc;
    Ort::AllocatedStringPtr inputName = session.GetInputNameAllocated(0, ort_alloc);
    Ort::AllocatedStringPtr outputName = session.GetOutputNameAllocated(0, ort_alloc);
    const std::array<const char*, 1> inputNames = { inputName.get() };
    const std::array<const char*, 1> outputNames = { outputName.get() };
    inputName.release();
    outputName.release();

    // run
    session.Run(runOptions, inputNames.data(), &inputTensor, 1, outputNames.data(), &outputTensor, 1);

    // copy output array to vector
    std::vector<float> output_vec;
    for (int i = 0; i < output.size(); ++i) {
        output_vec.push_back(output[i]);
    }

    return output_vec;
}

File: python/SAMImgEncoder.py

import json
import logging
import numpy as np
import torch
import onnxruntime
import matplotlib.pyplot as plt
from torch.nn import functional as F
from torchvision.transforms.functional import resize, to_pil_image  # type: ignore
from typing import Tuple
from copy import deepcopy
import cv2


class SAMImgEncoder:
    def __init__(self, config_path) -> None:
        with open(config_path) as f:
            config = json.load(f)
        
        # ResizeLongestSide: Proportional scaling the image.
        self.img_size = config['img_size']

        # set device
        self.device = config['device']
        if config['device'] == 'cuda' and torch.cuda.is_available():
            self.device = 'cuda'
        else:
            self.device = 'cpu'
        logging.info(f"Using device: {self.device}")

        # Preprocess: Normalize the image.
        self.pixel_mean = torch.Tensor([123.675, 116.28, 103.53]).view(-1, 1, 1).to(self.device)
        self.pixel_std = torch.Tensor([58.395, 57.12, 57.375]).view(-1, 1, 1).to(self.device)

        # Quantization
        self.qt = config['qt']
    
    def get_preprocess_shape(self, oldh: int, oldw: int, long_side_length: int) -> Tuple[int, int]:
        """
        Compute the output size given input size and target long side length.
        """
        scale = long_side_length * 1.0 / max(oldh, oldw)
        newh, neww = oldh * scale, oldw * scale
        neww = int(neww + 0.5)
        newh = int(newh + 0.5)
        return (newh, neww)
    
    def apply_image(self, image: np.ndarray) -> np.ndarray:
        """
        Expects a numpy array with shape HxWxC in uint8 format.
        """
        target_size = self.get_preprocess_shape(image.shape[0], image.shape[1], self.img_size)
        # print('apply_image.target_size:', target_size)
        # print("before_pil", image.flatten()[:100])
        # print('to_pil_image(image)', np.array(to_pil_image(image)).flatten()[:100], 
        #       np.array(to_pil_image(image)).shape) # (1200, 1800, 3)
        # print('resize(to_pil_image(image), target_size)', np.array(resize(to_pil_image(image), target_size)).flatten()[:100],
        #         np.array(resize(to_pil_image(image), target_size)).shape) # (683, 1024, 3) hwc

        return np.array(resize(to_pil_image(image), target_size))
    
    def apply_img_cv2(self, image: np.ndarray) -> np.ndarray:
        """
        Expects a numpy array with shape HxWxC in uint8 format.
        """
        target_size = self.get_preprocess_shape(image.shape[1], image.shape[0], self.img_size) # w,h
        # print('apply_image_cv2.target_size:', target_size)
        resized = cv2.resize(image, target_size, interpolation=cv2.INTER_LINEAR) # (1024, 683, 3) whc
        
        # print('apply_image_cv2.resized:', resized.flatten()[:100], resized.shape)
        return resized
    
    def apply_coords(self, coords: np.ndarray, original_size: Tuple[int, ...]) -> np.ndarray:
        """
        Expects a numpy array of length 2 in the final dimension. Requires the
        original image size in (H, W) format.
        """
        old_h, old_w = original_size
        new_h, new_w = self.get_preprocess_shape(
            original_size[0], original_size[1], self.img_size
        )
        coords = deepcopy(coords).astype(float)
        coords[..., 0] = coords[..., 0] * (new_w / old_w) # scale the coords from original image size to preprocessed image size.
        coords[..., 1] = coords[..., 1] * (new_h / old_h)
        return coords
    
    def encode_img(self, image: np.ndarray) -> np.ndarray:

        # resize the img, and convert to tensor
        input_image = self.apply_image(image) # note: image is RGB, HxWxC
        cv2_img = self.apply_img_cv2(image)
        print('PIL: ',input_image.flatten()[:100])
        print('cv2: ',cv2_img.flatten()[:100])
        # plt.imshow(input_image)
        # plt.show()
        # plt.imshow(cv2_img)
        # plt.show()
        # print(input_image.shape, cv2_img.shape)
        input_image_torch = torch.as_tensor(input_image, device=self.device)
        transformed_image = input_image_torch.permute(2, 0, 1).contiguous()[None, :, :, :] # HWC to CHW
        print('transformed_image:', transformed_image.flatten()[:100])
        assert (len(transformed_image.shape) == 4
                and transformed_image.shape[1] == 3
                and max(*transformed_image.shape[2:]) == self.img_size), "Invalid image shape before encode"

        # preprocess: normalize the image
        original_image_size = image.shape[:2]
        input_image = self.preprocess(transformed_image)
        print('self.preprocess:', input_image.shape)
        logging.info(f"Img size for Encoder: {original_image_size}")

        # select encoder or qt_encoder(int8)
        img_encoder_path = "../savedmodel/ImageEncoderViT.onnx"
        img_encoder_path_qt = "../savedmodel/ImageEncoderViT_qt.onnx"

        if self.device == 'cuda':
            providers=['CUDAExecutionProvider']
            logging.info("Using CUDA for ImageEncoder")
        else:
            providers=['CPUExecutionProvider']
            logging.info("Using CPU for ImageEncoder")

        if self.qt:
            self.img_encoderONNX = onnxruntime.InferenceSession(img_encoder_path_qt, providers=providers)
        else:
            self.img_encoderONNX = onnxruntime.InferenceSession(img_encoder_path, providers=providers)

        # encode the image
        input_feed = {'input': input_image.cpu().numpy()}
        features = self.img_encoderONNX.run(None, input_feed)
        print('features[0]:', features[0].shape)
        # features is a dict, so features[0] is the value.

        return features[0]

    def preprocess(self, x: torch.Tensor) -> torch.Tensor:
        """Normalize pixel values and pad to a square input."""
        # Normalize colors
        x = (x - self.pixel_mean) / self.pixel_std
        print("encode_img.prepocess.norm:", x.flatten()[:100])

        # Pad
        h, w = x.shape[-2:]
        padh = self.img_size - h
        padw = self.img_size - w
        x = F.pad(x, (0, padw, 0, padh))
        print("encode_img.prepocess.pad:", x[0])
        return x
